language/tools/move-cli/src/package.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use include_dir::{include_dir, Dir, DirEntry};
+use std::{
+    fs::{self, File},
+    io::Write,
+    path::Path,
+};
+use tempfile::tempdir;
+
+use move_lang::{compiled_unit::CompiledUnit, find_filenames, move_compile};
+use stdlib::{COMPILED_EXTENSION, MOVE_EXTENSION};
+use vm::file_format::CompiledModule;
+
+/// Directory name for the package source files under package/<name>
+const PKG_SOURCE_DIR: &str = "source_files";
+/// Directory name for the package binary files under package/<name>
+const PKG_BINARY_DIR: &str = "compiled";
+
+/// Content for the stdlib package
+const PACKAGE_STDLIB: MovePackage = MovePackage::pkg_stdlib();
+/// Content for the nursery package
+const PACKAGE_NURSERY: MovePackage = MovePackage::pkg_nursery();
+
+struct MovePackage {
+    /// Name of the package
+    name: &'static str,
+    /// The directory containing all the .move source files
+    source_dir: Dir<'static>,
+    /// The directory containing all the .mv binary files, if available
+    binary_dir: Option<Dir<'static>>,
+    /// Dependencies
+    deps: &'static [&'static MovePackage],
+}
+
+impl MovePackage {
+    /// Build the stdlib package statically
+    const fn pkg_stdlib() -> Self {
+        MovePackage {
+            name: "stdlib",
+            source_dir: include_dir!("../../stdlib/modules"),
+            binary_dir: Some(include_dir!("../../stdlib/compiled/stdlib")),
+            deps: &[],
+        }
+    }
+
+    /// Build the nursery package statically
+    const fn pkg_nursery() -> Self {
+        MovePackage {
+            name: "nursery",
+            source_dir: include_dir!("../../stdlib/nursery"),
+            binary_dir: None,
+            deps: &[&PACKAGE_STDLIB],
+        }
+    }
+
+    /// Lay down the source files to a given directory
+    pub fn prepare_source_files(&self, dir_path: &Path) -> Result<Vec<String>> {
+        let mut dep_dirs = vec![];
+
+        // bottom-up by preparing the dependencies first
+        for dep in self.deps {
+            dep_dirs.extend(dep.prepare_source_files(dir_path)?);
+        }
+
+        // prepare the output directory
+        let pkg_path = dir_path.join(self.name).join(PKG_SOURCE_DIR);
+        dep_dirs.push(pkg_path.clone().into_os_string().into_string().unwrap());
+
+        // if we have processed the package, shortcut the execution
+        if pkg_path.exists() {
+            return Ok(dep_dirs);
+        }
+
+        // write file content
+        fs::create_dir_all(&pkg_path)?;
+
+        let glob = format!("**/*.{}", MOVE_EXTENSION);
+        for entry in self.source_dir.find(&glob).unwrap() {
+            if let DirEntry::File(file) = entry {
+                let file_path = pkg_path.join(file.path());
+                let mut fp = File::create(file_path)?;
+                fp.write_all(file.contents())?;
+            } else {
```

Comment:
I really loathe if-let with an else, I think a `match` would do better here :)

Reply:
I think your opinion might be an outlier in Rust land, although I think I'd agree with it the case where you cared about more than one variant. `if let Some(foo) = bar { x } else { y }` is pretty natural.  But that does not apply to most uses of `Result` at all, where `match` is clearly warranted.

In any case, if you and the rest of the Move team feel strongly about this pattern, I'd suggest creating a coding guidelines doc with these kinds of things, otherwise you'll end up with inconsistent changes based on who the reviewers are.